### PR TITLE
feat(pvc, idiomatic): add idiomatic support to pvc object

### PR DESCRIPTION
**Pull request by mittachaitu**
Signed-off-by: mittachaitu <sai.chaithanya@mayadata.io>

**What this PR does**:
This PR enhances the persistent volume claim package to support for building pvc objects.  
**why we need it**:
It is required to create pvc objects using this package which built on top of k8s client-go.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Checklist:**
- [ ] Fixes #<issue number>
- [ ] Labelled this PR & related issue with `documentation` tag
- [ ] PR messages has document related information
- [ ] Labelled this PR & related issue with `breaking-changes` tag
- [ ] PR messages has breaking changes related information
- [ ] Labelled this PR & related issue with `requires-upgrade` tag
- [ ] PR messages has upgrade related information
- [ ] Commit has unit tests
- [ ] Commit has integration tests

### CODE REVIEW

File: pkg/kubernetes/persistentvolumeclaim/v1alpha1/kubernetes_test.go [hunk 1]
```diff
@@ -6,11 +6,12 @@ import (
 	"testing"
 
 	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
```
Comment (imazik) [inline pkg/kubernetes/persistentvolumeclaim/v1alpha1/kubernetes_test.go]: nit: `corev1`

File: pkg/kubernetes/persistentvolumeclaim/v1alpha1/kubernetes.go [hunk 2]
```diff
@@ -159,18 +169,27 @@ func (k *Kubeclient) Delete(name string, deleteOpts *metav1.DeleteOptions) error
...
+		return errors.Wrapf(err, "failed to delete pvc: '%s'", name)
 	}
 	return k.del(cli, k.namespace, name, deleteOpts)
 }
 
+// Create creates a pvc in specified namespace in kubernetes cluster
+func (k *Kubeclient) Create(pvc *corev1.PersistentVolumeClaim) (*corev1.PersistentVolumeClaim, error) {
+	cli, err := k.getClientSetOrCached()
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to create the persistent volume claim: '%v'", pvc)
```
[the remark this answers is not in the document]
Comment (AmitKumarDas) [reply pkg/kubernetes/persistentvolumeclaim/v1alpha1/kubernetes.go]: let us discuss to have the log pretty print pvc.
```go
return nil, errors.Wrapf(err, "failed to create pvc: failed to get clientset: %s", stringer.YAML("persistent volume claim", pvc))
```

File: pkg/kubernetes/persistentvolumeclaim/v1alpha1/persistentvolumeclaim.go [hunk 3]
```diff
@@ -180,3 +183,94 @@ func (b *ListBuilder) WithFilter(pred ...Predicate) *ListBuilder {
 	b.filters = append(b.filters, pred...)
 	return b
 }
+
+// NewPVC returns a PVC instance
+func NewPVC() *PVC {
```
Comment (AmitKumarDas) [inline pkg/kubernetes/persistentvolumeclaim/v1alpha1/persistentvolumeclaim.go]: This is not what idiomatic maya follows?
Why are we trying to deviate from our standard?

File: pkg/kubernetes/persistentvolumeclaim/v1alpha1/persistentvolumeclaim.go [hunk 4]
```diff
@@ -180,3 +183,94 @@ func (b *ListBuilder) WithFilter(pred ...Predicate) *ListBuilder {
...
+	return &PVC{Object: &corev1.PersistentVolumeClaim{}}
+}
+
+// WithName returns a PVC instance with name
+func (p *PVC) WithName(name string) *PVC {
```
Comment (AmitKumarDas) [inline pkg/kubernetes/persistentvolumeclaim/v1alpha1/persistentvolumeclaim.go]: This is not what idiomatic maya follows?
Why are we trying to deviate from our standard?

File: pkg/kubernetes/persistentvolumeclaim/v1alpha1/persistentvolumeclaim.go [hunk 5]
```diff
@@ -180,3 +183,94 @@ func (b *ListBuilder) WithFilter(pred ...Predicate) *ListBuilder {
...
+	if len(name) == 0 {
+		p.Err = errors.New("PVC name shouldn't be empty")
+		return p
+	}
+	p.Object.Name = name
+	return p
+}
+
+// WithNamespace returns a PVC instance with namespace
+func (p *PVC) WithNamespace(namespace string) *PVC {
```
Comment (AmitKumarDas) [inline pkg/kubernetes/persistentvolumeclaim/v1alpha1/persistentvolumeclaim.go]: This is not what idiomatic maya follows?
Why are we trying to deviate from our standard?

File: pkg/kubernetes/persistentvolumeclaim/v1alpha1/persistentvolumeclaim.go [hunk 6]
```diff
@@ -180,3 +183,94 @@ func (b *ListBuilder) WithFilter(pred ...Predicate) *ListBuilder {
...
+	}
+	if len(namespace) == 0 {
+		namespace = "default"
+	}
+	p.Object.Namespace = namespace
+	return p
+}
+
+// WithAnnotations returns a PVC instance with annotations
+func (p *PVC) WithAnnotations(annotations map[string]string) *PVC {
```
Comment (AmitKumarDas) [inline pkg/kubernetes/persistentvolumeclaim/v1alpha1/persistentvolumeclaim.go]: This is not what idiomatic maya follows?
Why are we trying to deviate from our standard?

File: pkg/kubernetes/persistentvolumeclaim/v1alpha1/persistentvolumeclaim.go [hunk 7]
```diff
@@ -180,3 +183,94 @@ func (b *ListBuilder) WithFilter(pred ...Predicate) *ListBuilder {
...
+	if p.Err != nil {
+		return p
+	}
+	p.Object.Annotations = annotations
+	return p
+}
+
+// WithLabels returns a PVC instance with labels
+func (p *PVC) WithLabels(labels map[string]string) *PVC {
```
Comment (AmitKumarDas) [inline pkg/kubernetes/persistentvolumeclaim/v1alpha1/persistentvolumeclaim.go]: This is not what idiomatic maya follows?
Why are we trying to deviate from our standard?

File: pkg/kubernetes/persistentvolumeclaim/v1alpha1/persistentvolumeclaim.go [hunk 8]
```diff
@@ -180,3 +183,94 @@ func (b *ListBuilder) WithFilter(pred ...Predicate) *ListBuilder {
...
+	if p.Err != nil {
+		return p
+	}
+	p.Object.Labels = labels
+	return p
+}
+
+// WithStorageClass returns a PVC instance with storageclass
+func (p *PVC) WithStorageClass(scName string) *PVC {
```
Comment (AmitKumarDas) [inline pkg/kubernetes/persistentvolumeclaim/v1alpha1/persistentvolumeclaim.go]: This is not what idiomatic maya follows?
Why are we trying to deviate from our standard?

File: pkg/kubernetes/persistentvolumeclaim/v1alpha1/persistentvolumeclaim.go [hunk 9]
```diff
@@ -180,3 +183,94 @@ func (b *ListBuilder) WithFilter(pred ...Predicate) *ListBuilder {
...
+	if len(scName) == 0 {
+		p.Err = errors.New("PVC storageclass name shouldn't be empty")
+		return p
+	}
+	p.Object.Spec.StorageClassName = &scName
+	return p
+}
+
+// WithAccessModes returns a PVC instance with annotations
+func (p *PVC) WithAccessModes(accessMode []corev1.PersistentVolumeAccessMode) *PVC {
```
Comment (AmitKumarDas) [inline pkg/kubernetes/persistentvolumeclaim/v1alpha1/persistentvolumeclaim.go]: This is not what idiomatic maya follows?
Why are we trying to deviate from our standard?

File: pkg/kubernetes/persistentvolumeclaim/v1alpha1/persistentvolumeclaim.go [hunk 10]
```diff
@@ -180,3 +183,94 @@ func (b *ListBuilder) WithFilter(pred ...Predicate) *ListBuilder {
...
+	if len(accessMode) == 0 {
+		p.Err = errors.New("PVC accessMode shouldn't be empty")
+		return p
+	}
+	p.Object.Spec.AccessModes = accessMode
+	return p
+}
+
+// WithCapacity returns a PVC instance with capacity
+func (p *PVC) WithCapacity(capacity string) *PVC {
```
Comment (AmitKumarDas) [inline pkg/kubernetes/persistentvolumeclaim/v1alpha1/persistentvolumeclaim.go]: This is not what idiomatic maya follows?
Why are we trying to deviate from our standard?

File: pkg/kubernetes/persistentvolumeclaim/v1alpha1/kubernetes.go [hunk 11]
```diff
@@ -116,39 +126,39 @@ func NewKubeClient(opts ...KubeclientBuildOption) *Kubeclient {
...
-func (k *Kubeclient) Get(name string, opts metav1.GetOptions) (*v1.PersistentVolumeClaim, error) {
+func (k *Kubeclient) Get(name string, opts metav1.GetOptions) (*corev1.PersistentVolumeClaim, error) {
 	if strings.TrimSpace(name) == "" {
 		return nil, errors.New("failed to get pvc: missing pvc name")
 	}
-	cli, err := k.getClientOrCached()
+	cli, err := k.getClientSetOrCached()
 	if err != nil {
-		return nil, err
+		return nil, errors.Wrapf(err, "failed to get the pvc: '%s'", name)
```
[the remark this answers is not in the document]
Comment (mittachaitu) [reply pkg/kubernetes/persistentvolumeclaim/v1alpha1/kubernetes.go]: Done

File: pkg/kubernetes/persistentvolumeclaim/v1alpha1/kubernetes.go [hunk 12]
```diff
@@ -159,18 +169,27 @@ func (k *Kubeclient) Delete(name string, deleteOpts *metav1.DeleteOptions) error
 	if strings.TrimSpace(name) == "" {
 		return errors.New("failed to delete pvc: missing pvc name")
 	}
-	cli, err := k.getClientOrCached()
+	cli, err := k.getClientSetOrCached()
 	if err != nil {
-		return err
+		return errors.Wrapf(err, "failed to delete pvc: '%s'", name)
```
[the remark this answers is not in the document]
Comment (mittachaitu) [reply pkg/kubernetes/persistentvolumeclaim/v1alpha1/kubernetes.go]: Done

File: pkg/kubernetes/persistentvolumeclaim/v1alpha1/kubernetes.go [hunk 13]
```diff
@@ -116,39 +126,39 @@ func NewKubeClient(opts ...KubeclientBuildOption) *Kubeclient {
...
 
 // List returns a list of pvc
 // instances present in kubernetes cluster
-func (k *Kubeclient) List(opts metav1.ListOptions) (*v1.PersistentVolumeClaimList, error) {
-	cli, err := k.getClientOrCached()
+func (k *Kubeclient) List(opts metav1.ListOptions) (*corev1.PersistentVolumeClaimList, error) {
+	cli, err := k.getClientSetOrCached()
 	if err != nil {
-		return nil, err
+		return nil, errors.Wrapf(err, "failed to list pvc")
```
[the remark this answers is not in the document]
Comment (mittachaitu) [reply pkg/kubernetes/persistentvolumeclaim/v1alpha1/kubernetes.go]: Done

File: pkg/kubernetes/persistentvolumeclaim/v1alpha1/kubernetes.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (mittachaitu) [reply pkg/kubernetes/persistentvolumeclaim/v1alpha1/kubernetes.go]: Done

File: pkg/kubernetes/persistentvolumeclaim/v1alpha1/kubernetes.go [hunk 14]
```diff
@@ -159,18 +169,27 @@ func (k *Kubeclient) Delete(name string, deleteOpts *metav1.DeleteOptions) error
...
+	return k.create(cli, k.namespace, pvc)
+}
+
 // DeleteCollection deletes a collection of pvc objects.
 func (k *Kubeclient) DeleteCollection(listOpts metav1.ListOptions, deleteOpts *metav1.DeleteOptions) error {
-	cli, err := k.getClientOrCached()
+	cli, err := k.getClientSetOrCached()
 	if err != nil {
-		return err
+		return errors.Wrapf(err, "failed to delete the collection of pvcs")
```
[the remark this answers is not in the document]
Comment (mittachaitu) [reply pkg/kubernetes/persistentvolumeclaim/v1alpha1/kubernetes.go]: Done

File: pkg/kubernetes/persistentvolumeclaim/v1alpha1/persistentvolumeclaim.go [hunk 15]
```diff
@@ -3,14 +3,17 @@ package v1alpha1
...
+	"k8s.io/apimachinery/pkg/api/resource"
 )
 
 // PVC is a wrapper over persistentvolumeclaim api
 // object. It provides build, validations and other common
 // logic to be used by various feature specific callers.
 type PVC struct {
-	object *corev1.PersistentVolumeClaim
+	Object *corev1.PersistentVolumeClaim
+	Err    error
```
[the remark this answers is not in the document]
Comment (mittachaitu) [reply pkg/kubernetes/persistentvolumeclaim/v1alpha1/persistentvolumeclaim.go]: Done

File: pkg/kubernetes/persistentvolumeclaim/v1alpha1/persistentvolumeclaim.go (hunk 3 above)
Comment (mittachaitu) [reply pkg/kubernetes/persistentvolumeclaim/v1alpha1/persistentvolumeclaim.go]: Refactored the code

File: pkg/kubernetes/persistentvolumeclaim/v1alpha1/persistentvolumeclaim.go (hunk 4 above)
Comment (mittachaitu) [reply pkg/kubernetes/persistentvolumeclaim/v1alpha1/persistentvolumeclaim.go]: Refactored the code

File: pkg/kubernetes/persistentvolumeclaim/v1alpha1/persistentvolumeclaim.go (hunk 5 above)
Comment (mittachaitu) [reply pkg/kubernetes/persistentvolumeclaim/v1alpha1/persistentvolumeclaim.go]: Refactored the code

File: pkg/kubernetes/persistentvolumeclaim/v1alpha1/persistentvolumeclaim.go (hunk 6 above)
Comment (mittachaitu) [reply pkg/kubernetes/persistentvolumeclaim/v1alpha1/persistentvolumeclaim.go]: Refactored the code

File: pkg/kubernetes/persistentvolumeclaim/v1alpha1/persistentvolumeclaim.go (hunk 7 above)
Comment (mittachaitu) [reply pkg/kubernetes/persistentvolumeclaim/v1alpha1/persistentvolumeclaim.go]: Refactored the code

File: pkg/kubernetes/persistentvolumeclaim/v1alpha1/persistentvolumeclaim.go (hunk 8 above)
Comment (mittachaitu) [reply pkg/kubernetes/persistentvolumeclaim/v1alpha1/persistentvolumeclaim.go]: Refactored the code

File: pkg/kubernetes/persistentvolumeclaim/v1alpha1/persistentvolumeclaim.go (hunk 9 above)
Comment (mittachaitu) [reply pkg/kubernetes/persistentvolumeclaim/v1alpha1/persistentvolumeclaim.go]: Refactored the code

File: pkg/kubernetes/persistentvolumeclaim/v1alpha1/persistentvolumeclaim.go (hunk 10 above)
Comment (mittachaitu) [reply pkg/kubernetes/persistentvolumeclaim/v1alpha1/persistentvolumeclaim.go]: Refactored the code

File: pkg/kubernetes/persistentvolumeclaim/v1alpha1/kubernetes.go [hunk 16]
```diff
@@ -116,39 +127,39 @@ func NewKubeClient(opts ...KubeclientBuildOption) *Kubeclient {
...
-func (k *Kubeclient) Get(name string, opts metav1.GetOptions) (*v1.PersistentVolumeClaim, error) {
+func (k *Kubeclient) Get(name string, opts metav1.GetOptions) (*corev1.PersistentVolumeClaim, error) {
 	if strings.TrimSpace(name) == "" {
 		return nil, errors.New("failed to get pvc: missing pvc name")
 	}
-	cli, err := k.getClientOrCached()
+	cli, err := k.getClientSetOrCached()
 	if err != nil {
-		return nil, err
+		return nil, errors.Wrapf(err, "failed to get pvc {%s}: failed to get clientset", name)
```
Comment (prateekpandey14) [inline pkg/kubernetes/persistentvolumeclaim/v1alpha1/kubernetes.go]: nit: Let's remove the redundant error ` failed to get clientset` as `getClientSet` method is already returns the same error, which will eventually appended to the err.

File: pkg/kubernetes/persistentvolumeclaim/v1alpha1/kubernetes.go [hunk 17]
```diff
@@ -116,39 +127,39 @@ func NewKubeClient(opts ...KubeclientBuildOption) *Kubeclient {
...
 
 // List returns a list of pvc
 // instances present in kubernetes cluster
-func (k *Kubeclient) List(opts metav1.ListOptions) (*v1.PersistentVolumeClaimList, error) {
-	cli, err := k.getClientOrCached()
+func (k *Kubeclient) List(opts metav1.ListOptions) (*corev1.PersistentVolumeClaimList, error) {
+	cli, err := k.getClientSetOrCached()
 	if err != nil {
-		return nil, err
+		return nil, errors.Wrapf(err, "failed to list pvc: failed to get clientset")
```
Comment (prateekpandey14) [inline pkg/kubernetes/persistentvolumeclaim/v1alpha1/kubernetes.go]: Same as above comment... thanks

File: pkg/kubernetes/persistentvolumeclaim/v1alpha1/kubernetes.go (hunk 16 above)
Comment (mittachaitu) [reply pkg/kubernetes/persistentvolumeclaim/v1alpha1/kubernetes.go]: Done

File: pkg/kubernetes/persistentvolumeclaim/v1alpha1/buildlist.go [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+// TODO add validations and error checks
+func (b *ListBuilder) Build() (*PVCList, error) {
+	return b.list, nil
+}
+
+// ListBuilderForAPIObjects builds the ListBuilder object based on PVC api list
+func ListBuilderForAPIObjects(pvcs *corev1.PersistentVolumeClaimList) *ListBuilder {
+	b := &ListBuilder{list: &PVCList{}}
+	if pvcs == nil {
+		return b
```
Comment (AmitKumarDas) [inline pkg/kubernetes/persistentvolumeclaim/v1alpha1/buildlist.go]: add error here
```go
"failed to build pvc list: missing api list",
```

Comment (mittachaitu) [reply pkg/kubernetes/persistentvolumeclaim/v1alpha1/buildlist.go]: Done

File: pkg/kubernetes/persistentvolumeclaim/v1alpha1/build.go [hunk 19]
```diff
@@ -0,0 +1,112 @@
...
+		b.err = errors.New("failed to build PVC object: missing accessmodes types")
+		return b
+	}
+	b.Pvc.object.Spec.AccessModes = accessMode
+	return b
+}
+
+// WithCapacity sets the Capacity field in PVC with provided arguments
+func (b *Builder) WithCapacity(capacity string) *Builder {
+	if b.err != nil {
```
[the remark this answers is not in the document]
Comment (mittachaitu) [reply pkg/kubernetes/persistentvolumeclaim/v1alpha1/build.go]: Done

File: pkg/kubernetes/persistentvolumeclaim/v1alpha1/build.go [hunk 20]
```diff
@@ -0,0 +1,112 @@
...
+	return b
+}
+
+// WithAccessModes sets the AccessMode field in PVC with provided arguments
+func (b *Builder) WithAccessModes(accessMode []corev1.PersistentVolumeAccessMode) *Builder {
+	if b.err != nil {
+		return b
+	}
+	if len(accessMode) == 0 {
+		b.err = errors.New("failed to build PVC object: missing accessmodes types")
```
[the remark this answers is not in the document]
Comment (mittachaitu) [reply pkg/kubernetes/persistentvolumeclaim/v1alpha1/build.go]: Done

File: pkg/kubernetes/persistentvolumeclaim/v1alpha1/build.go [hunk 21]
```diff
@@ -0,0 +1,112 @@
...
+		b.err = errors.New("failed to build PVC object: missing storageclass name")
+		return b
+	}
+	b.Pvc.object.Spec.StorageClassName = &scName
+	return b
+}
+
+// WithAccessModes sets the AccessMode field in PVC with provided arguments
+func (b *Builder) WithAccessModes(accessMode []corev1.PersistentVolumeAccessMode) *Builder {
+	if b.err != nil {
```
[the remark this answers is not in the document]
Comment (mittachaitu) [reply pkg/kubernetes/persistentvolumeclaim/v1alpha1/build.go]: Removed check

File: pkg/kubernetes/persistentvolumeclaim/v1alpha1/build.go [hunk 22]
```diff
@@ -0,0 +1,112 @@
...
+		return b
+	}
+	b.Pvc.object.Labels = labels
+	return b
+}
+
+// WithStorageClass sets the StorageClass field of PVC with provided arguments
+func (b *Builder) WithStorageClass(scName string) *Builder {
+	if b.err != nil {
```
[the remark this answers is not in the document]
Comment (mittachaitu) [reply pkg/kubernetes/persistentvolumeclaim/v1alpha1/build.go]: Removed check

File: pkg/kubernetes/persistentvolumeclaim/v1alpha1/build.go [hunk 23]
```diff
@@ -0,0 +1,112 @@
...
+	if len(namespace) == 0 {
+		namespace = "default"
+	}
+	b.Pvc.object.Namespace = namespace
+	return b
+}
+
+// WithAnnotations sets the Annotations field of PVC with provided arguments
+func (b *Builder) WithAnnotations(annotations map[string]string) *Builder {
+	if b.err != nil {
```
[the remark this answers is not in the document]
Comment (mittachaitu) [reply pkg/kubernetes/persistentvolumeclaim/v1alpha1/build.go]: Removed check

File: pkg/kubernetes/persistentvolumeclaim/v1alpha1/build.go [hunk 24]
```diff
@@ -0,0 +1,112 @@
...
+		b.err = errors.New("failed to build PVC object: missing PVC name")
+		return b
+	}
+	b.Pvc.object.Name = name
+	return b
+}
+
+// WithNamespace sets the Namespace field of PVC provided arguments
+func (b *Builder) WithNamespace(namespace string) *Builder {
+	if b.err != nil {
```
[the remark this answers is not in the document]
Comment (mittachaitu) [reply pkg/kubernetes/persistentvolumeclaim/v1alpha1/build.go]: Done

File: pkg/kubernetes/persistentvolumeclaim/v1alpha1/build.go [hunk 25]
```diff
@@ -0,0 +1,112 @@
...
+import (
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+)
+
+// Builder is the builder object for PVC
+type Builder struct {
+	Pvc *PVC
+	err error
```
[the remark this answers is not in the document]
Comment (mittachaitu) [reply pkg/kubernetes/persistentvolumeclaim/v1alpha1/build.go]: Done

File: pkg/kubernetes/persistentvolumeclaim/v1alpha1/build.go [hunk 26]
```diff
@@ -0,0 +1,112 @@
...
+
+import (
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+)
+
+// Builder is the builder object for PVC
+type Builder struct {
+	Pvc *PVC
```
[the remark this answers is not in the document]
Comment (mittachaitu) [reply pkg/kubernetes/persistentvolumeclaim/v1alpha1/build.go]: Done

File: pkg/kubernetes/persistentvolumeclaim/v1alpha1/build.go [hunk 27]
```diff
@@ -0,0 +1,115 @@
...
+	}
+	b.pvc.object.Spec.AccessModes = accessMode
+	return b
+}
+
+// WithCapacity sets the Capacity field in PVC with provided arguments
+func (b *Builder) WithCapacity(capacity string) *Builder {
+	resCapacity, err := resource.ParseQuantity(capacity)
+	if err != nil {
+		b.errs = append(b.errs, errors.Errorf("failed to build PVC object: failed to parse capacity {%s}", capacity))
```
Comment (AmitKumarDas) [inline pkg/kubernetes/persistentvolumeclaim/v1alpha1/build.go]: We are missing the actual error.
errors.Wrapf() should be used.

File: pkg/kubernetes/persistentvolumeclaim/v1alpha1/buildlist.go [hunk 28]
```diff
@@ -0,0 +1,103 @@
...
+}
+
+// Len returns the number of items present
+// in the PVCList of a builder
+func (b *ListBuilder) Len() int {
+	return len(b.list.items)
+}
+
+// APIList builds core API PVC list using listbuilder
+func (b *ListBuilder) APIList() (*corev1.PersistentVolumeClaimList, []error) {
```
Comment (AmitKumarDas) [inline pkg/kubernetes/persistentvolumeclaim/v1alpha1/buildlist.go]: The return type will always be `error`. In other words here it should be (*corev1.PersistentVolumeClaimList, error)
```go
return nil, errors.Errorf("%+v", errs)
```

File: pkg/kubernetes/persistentvolumeclaim/v1alpha1/buildlist.go [hunk 29]
```diff
@@ -0,0 +1,103 @@
...
+	l, errs := b.Build()
+	if len(errs) > 0 {
+		return nil, errs
+	}
+	if l == nil {
```
Comment (AmitKumarDas) [inline pkg/kubernetes/persistentvolumeclaim/v1alpha1/buildlist.go]: Let us remove this check. This is not required.

File: pkg/kubernetes/persistentvolumeclaim/v1alpha1/buildlist.go [hunk 30]
```diff
@@ -0,0 +1,103 @@
...
+			filtered.items = append(filtered.items, pvc)
+		}
+	}
+	return filtered
+}
+
+// Len returns the number of items present
+// in the PVCList of a builder
+func (b *ListBuilder) Len() int {
+	return len(b.list.items)
```
Comment (AmitKumarDas) [inline pkg/kubernetes/persistentvolumeclaim/v1alpha1/buildlist.go]: It should call the Build() as well.
Len() can return (int, error)

File: pkg/kubernetes/persistentvolumeclaim/v1alpha1/buildlist.go [hunk 31]
```diff
@@ -0,0 +1,103 @@
...
+func NewListBuilder() *ListBuilder {
+	return &ListBuilder{list: &PVCList{}}
+}
+
+// Build returns the final instance of patch
+func (b *ListBuilder) Build() (*PVCList, []error) {
+	if len(b.errs) > 0 {
+		return nil, b.errs
+	}
+	return b.list, nil
```
Comment (AmitKumarDas) [inline pkg/kubernetes/persistentvolumeclaim/v1alpha1/buildlist.go]: Build must make use of filters as well.
Build(), List(), APIList() & Len() -- all are final functions of Builder.
List(), APIList() & Len() should make use of Build().
Build() should make use of all its properties including `filters`.

File: pkg/kubernetes/persistentvolumeclaim/v1alpha1/build.go (hunk 27 above)
Comment (mittachaitu) [reply pkg/kubernetes/persistentvolumeclaim/v1alpha1/build.go]: Updated with errors.Wrapf()

File: pkg/kubernetes/persistentvolumeclaim/v1alpha1/buildlist.go (hunk 28 above)
Comment (mittachaitu) [reply pkg/kubernetes/persistentvolumeclaim/v1alpha1/buildlist.go]: List() is internally called by APIList to apply fillers and get filterdlist.
List() will return the  `return nil, errors.Errorf("%+v", errs)` if ListBuilder has any error

File: pkg/kubernetes/persistentvolumeclaim/v1alpha1/buildlist.go (hunk 29 above)
Comment (mittachaitu) [reply pkg/kubernetes/persistentvolumeclaim/v1alpha1/buildlist.go]: Done

File: pkg/kubernetes/persistentvolumeclaim/v1alpha1/buildlist.go (hunk 30 above)
Comment (mittachaitu) [reply pkg/kubernetes/persistentvolumeclaim/v1alpha1/buildlist.go]: same as above comment(Len() will internally invoke List())

File: pkg/kubernetes/persistentvolumeclaim/v1alpha1/buildlist.go (hunk 31 above)
Comment (mittachaitu) [reply pkg/kubernetes/persistentvolumeclaim/v1alpha1/buildlist.go]: Done

File: integration-tests/sts/sts_test.go [hunk 32]
```diff
@@ -131,11 +131,11 @@ var _ = Describe("StatefulSet", func() {
 				NewKubeClient(pvc.WithNamespace(stsNamespace)).
 				List(metav1.ListOptions{LabelSelector: stsApplicationLabel})
 			Expect(err).ShouldNot(HaveOccurred())
-			return pvc.
+			pvcCount, _ := pvc.
 				ListBuilderForAPIObjects(pvcs).
 				WithFilter(pvc.IsBound()).
-				List().
 				Len()
+			return pvcCount
```
Comment (AmitKumarDas) [inline integration-tests/sts/sts_test.go]: Should we check if err should not have occurred

**Pull request merged**